File: applications/rtkextractphasesignal/rtkextractphasesignal.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkextractphasesignal_ggo.h"
#include "rtkMacro.h"

#include "rtkExtractPhaseImageFilter.h"

#include <itkImageFileReader.h>
#include <fstream>

namespace rtk
{
template <class TSignalType>
void
WriteSignalToTextFile(TSignalType * sig, const std::string & fileName)
{
  std::ofstream                              ofs(fileName.c_str());
  itk::ImageRegionConstIterator<TSignalType> it(sig, sig->GetLargestPossibleRegion());
  for (it.GoToBegin(); !it.IsAtEnd(); ++it)
  {
    ofs << it.Get() << std::endl;
  }
  ofs.close();
}
} // namespace rtk

int
main(int argc, char * argv[])
{
  GGO(rtkextractphasesignal, args_info);

  constexpr unsigned int Dimension = 1;

  using ImageType = itk::Image<double, Dimension>;

  // Read
  ImageType::Pointer signal;
  signal = itk::ReadImage<ImageType>(args_info.input_arg);

  // Process phase signal if required
  using PhaseFilter = rtk::ExtractPhaseImageFilter<ImageType>;
  auto phase = PhaseFilter::New();
  phase->SetInput(signal);
  phase->SetMovingAverageSize(args_info.movavg_arg);
  phase->SetUnsharpMaskSize(args_info.unsharp_arg);
  phase->SetModel((PhaseFilter::ModelType)args_info.model_arg);
  TRY_AND_EXIT_ON_ITK_EXCEPTION(phase->Update())

  // Write output phase
  rtk::WriteSignalToTextFile(phase->GetOutput(), args_info.output_arg);

  return EXIT_SUCCESS;
}

File: applications/rtkextractphasesignal/rtkextractphasesignal.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk


def write_signal_to_text_file(signal_image, filename):

    # Convert ITK image to numpy array
    signal_array = itk.array_from_image(signal_image)

    # Write to text file
    with open(filename, 'w') as f:
        for value in signal_array.flatten():
            f.write(f"{value}\n")


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Extracts the phase from a signal."
    )

    # General options
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--input", "-i", help="Input signal", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name for the Hilbert phase signal", type=str, required=True
    )
    parser.add_argument(
        "--movavg", help="Moving average size applied before phase extraction", type=int, default=1
    )
    parser.add_argument(
        "--unsharp", help="Unsharp mask size applied before phase extraction", type=int, default=55
    )
    parser.add_argument(
        "--model",
        help="Phase model",
        choices=["LOCAL_PHASE", "LINEAR_BETWEEN_MINIMA", "LINEAR_BETWEEN_MAXIMA"],
        default="LINEAR_BETWEEN_MINIMA"
    )

    return parser


def process(args_info: argparse.Namespace):
    # Define pixel type and dimension
    PixelType = itk.D  # double
    Dimension = 1
    ImageType = itk.Image[PixelType, Dimension]

    # Read input signal
    if args_info.verbose:
        print(f"Reading input signal from {args_info.input}...")

    reader = itk.ImageFileReader[ImageType].New()
    reader.SetFileName(args_info.input)
    reader.Update()
    signal = reader.GetOutput()

    # Process phase signal if required
    PhaseFilter = rtk.ExtractPhaseImageFilter[ImageType]
    phase = PhaseFilter.New()
    phase.SetInput(signal)
    phase.SetMovingAverageSize(args_info.movavg)
    phase.SetUnsharpMaskSize(args_info.unsharp)

    # Map string to enum integer value
    model_values = {
        "LOCAL_PHASE": 0,
        "LINEAR_BETWEEN_MINIMA": 1,
        "LINEAR_BETWEEN_MAXIMA": 2
    }
    phase.SetModel(model_values[args_info.model])

    # Write output phase signal
    if args_info.verbose:
        print(f"Writing phase signal to {args_info.output}...")

    write_signal_to_text_file(phase.GetOutput(), args_info.output)

    if args_info.verbose:
        print("Phase extraction completed successfully.")


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
